src/bridge/AdminRoomHandler.ts

```diff
@@ -484,7 +484,7 @@ export class AdminRoomHandler {
                     `Username is longer than the maximum permitted by the bridge (${SANE_USERNAME_LENGTH}).`
                 );
             }
-            else if (IdentGenerator.sanitiseUsername(username) !== username) {
+            else if (!IdentGenerator.validateUsername(username)) {
```

Comment:
`validateUsername` is way too restrictive, because appservice's `!username` command is used primarily to configure the account name (required to authenticate), not just the username (which people rarely care about). For example, `validateUsername` is very strict about the first character, but account names don't have such restrictions.

Additionally, they have conflicting requirements: `IdentGenerator.validateUsername` can (and should) be overly restrictive when validating usernames because at worst the username will be mangled (like I said, no one cares about it); but `!username` really should not risk rejecting valid account names because prevents users from authenticating.

Reply:
@progval I'm not sure I understand. PR adds a comment inside `validateUsername` that the rules about the first character were taken [from there](https://github.com/solanum-ircd/solanum/blob/78825899cd6b68037fd1b3c6c77afb7d10c4774e/ircd/s_user.c#L914-L927). Could you please clarify what exactly is wrong with this rule?

> but `!username` really should not risk rejecting valid account names because prevents users from authenticating

I just want to point out that it has always been rejecting valid accounts, so at least some improvement in that regard is definitely wanted.

Reply:
The "username" passed as argument to `!username` is a [SASL PLAIN authcid, which can be any UTF8 string not containing a null char](https://www.rfc-editor.org/rfc/rfc4616#section-2).

The "username" that `IdentGenerator.validateUsername` validates is what is passed to the `USER` command, as a substitute for the ident when the client doesn't run an identd, and the grammar is unspecified.

It makes no sense to use the same code for both just because they are both informally called "username".

Reply:
Re-reading the code, I just noticed `IdentGenerator.validateUsername` is only used to validate the SASL username, so it shouldn't be `IdentGenerator` at all; and the quoted grammar is irrelevant because it's the grammar for `USER` on solanum, not for SASL.

Instead it should simply check there is no null character in the string.

Reply:
Oh, I think I see now, so you are saying that `handleUsername()` function being modified in the PR is called from two different locations: from `!username` and from some other `USER` command, right? So, the correct fix would be to only change username sanitization for the `!username` case.

Reply:
> Re-reading the code, I just noticed `IdentGenerator.validateUsername` is only used to validate the SASL username, so it shouldn't be `IdentGenerator` at all; and the quoted grammar is irrelevant because it's the grammar for `USER` on solanum, not for SASL.

Sorry, the page didn't update with a new comment, so I didn't see it when sent the previous one. So, is it the wrong code to modify for the `!username` sanitization problem?

Reply:
yes, it's the wrong code. It makes the bug a little less of a problem by being less strict (as you pointed out), but it's still needlessly strict.